daemon/emer-aggregate-tally.c

```diff
@@ -0,0 +1,373 @@
+                            GVariant  *aggregate_key,
+                            GVariant  *payload,
+                            guint32    counter,
+                            gint64     monotonic_time_us)
+{
+  return g_variant_new ("(vvuumv)",
+                        event_id,
+                        aggregate_key,
+                        unix_user_id,
+                        counter,
+                        payload);
+}
+
+static gchar *
+get_tally_path_from_aggregate_timer (EmerAggregateTally *self,
+                                     guint32             unix_user_id,
+                                     GVariant           *event_id,
+                                     GVariant           *aggregate_key,
+                                     const char         *date)
+{
+  g_autofree gchar *aggregate_key_str = NULL;
+  g_autofree gchar *event_id_str = NULL;
+  g_autofree gchar *filename = NULL;
+
+  aggregate_key_str = g_variant_print (aggregate_key, FALSE);
+  event_id_str = g_variant_print (event_id, FALSE);
+  filename = g_strdup_printf ("%s-%u-%s",
+                              aggregate_key_str,
+                              unix_user_id,
+                              event_id_str);
+
+  return g_build_filename (self->persistent_cache_directory,
+                           "aggregate-timers",
+                           date,
+                           filename,
+                           NULL);
+}
+
+static guint32
+find_previous_aggregate_timer_counter (EmerAggregateTally *self,
+                                       guint32             unix_user_id,
+                                       GVariant           *event_id,
+                                       GVariant           *aggregate_key,
+                                       const char         *date)
+{
+  g_autofree gchar *tally_path = NULL;
+  g_autofree gchar *contents = NULL;
+  g_autoptr(GVariant) variant = NULL;
+  g_autoptr(GVariant) unused_event_id = NULL;
+  g_autoptr(GVariant) unused_aggregate_key = NULL;
+  g_autoptr(GVariant) unused_payload = NULL;
+  g_autoptr(GError) error = NULL;
+  guint32 previous_counter;
+  guint32 unused_unix_user_id;
+  gsize size;
+
+  tally_path = get_tally_path_from_aggregate_timer (self,
+                                                    unix_user_id,
+                                                    event_id,
+                                                    aggregate_key,
+                                                    date);
+
+  if (!g_file_test (tally_path, G_FILE_TEST_EXISTS))
+    return 0;
+
+  g_file_get_contents (tally_path, &contents, &size, &error);
+  if (error)
+    return 0;
+
+  variant = g_variant_new_from_data (G_VARIANT_TYPE ("(vvuumv)"),
+                                     contents, size, FALSE, NULL, NULL);
+  if (error)
+    return 0;
+
+  if (!g_variant_check_format_string (variant, "(vvuumv)", FALSE))
+    return 0;
+
+  g_variant_get (variant,
+                 "(vvvuumv)",
```

Comment:
This format string is wrong, it has too many `v`s. Perhaps put a `#define` somewhere?

You can either pass `NULL` to ignore a value, or use `g_variant_get_child` to pick out a single field by index. No need for the `unused_*` fields.

Reply:
TIL, thanks